flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java

```diff
@@ -0,0 +1,145 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * A {@link RunnableWithException} executor that schedules a next attempt upon timeout based on
+ * {@link RetryPolicy}. Aimed to curb tail latencies
+ */
+class RetryingExecutor implements AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(RetryingExecutor.class);
+
+    private final ScheduledExecutorService scheduler;
+
+    RetryingExecutor(int nThreads) {
+        this(SchedulerFactory.create(nThreads, "ChangelogRetryScheduler", LOG));
+    }
+
+    RetryingExecutor(ScheduledExecutorService scheduler) {
+        this.scheduler = scheduler;
+    }
+
+    void execute(RetryPolicy retryPolicy, RunnableWithException action) {
+        LOG.debug("execute with retryPolicy: {}", retryPolicy);
+        RetriableTask task = new RetriableTask(action, retryPolicy, scheduler);
+        scheduler.submit(task);
+    }
+
+    @Override
+    public void close() throws Exception {
+        LOG.debug("close");
+        scheduler.shutdownNow();
+        if (!scheduler.awaitTermination(1, TimeUnit.SECONDS)) {
+            LOG.warn("Unable to cleanly shutdown executorService in 1s");
+        }
+    }
+
+    private static final class RetriableTask implements Runnable {
+        private final RunnableWithException runnable;
+        private final ScheduledExecutorService executorService;
+        private final int current;
+        private final RetryPolicy retryPolicy;
+        private final AtomicBoolean actionCompleted;
+        private final AtomicBoolean attemptCompleted = new AtomicBoolean(false);
+
+        RetriableTask(
+                RunnableWithException runnable,
+                RetryPolicy retryPolicy,
+                ScheduledExecutorService executorService) {
+            this(1, new AtomicBoolean(false), runnable, retryPolicy, executorService);
+        }
+
+        private RetriableTask(
+                int current,
+                AtomicBoolean actionCompleted,
+                RunnableWithException runnable,
+                RetryPolicy retryPolicy,
+                ScheduledExecutorService executorService) {
+            this.current = current;
+            this.runnable = runnable;
+            this.retryPolicy = retryPolicy;
+            this.executorService = executorService;
+            this.actionCompleted = actionCompleted;
+        }
+
+        @Override
+        public void run() {
+            if (!actionCompleted.get()) {
+                Optional<ScheduledFuture<?>> timeoutFuture = scheduleTimeout();
+                try {
+                    runnable.run();
```

Comment:
How are you solving a (race?) condition when two attempts complete? Do you have a requirement that `RunnableWithException runnable` must be idempotent and you just ignore that issue?

Reply:
Yes, `runnable` must be idempotent, as multiple instances of it can be executed concurrently.
I'll add this requirement to the javadoc, thanks for pointing out.

Reply:
I've also found an issue in `UploadTask`. It was idempotent already (thanks to checks in writer); but not thread-safe (because `finished` wasn't thread safe). I've fixed it in 9bba8b004352b54e8d99277146708ed7525bff56.

Reply:
Can you introduce a separate interface for that? `IdempotentRunnable` or `IdempotentRunnableWithException`? I believe this is so important requirement that it should be marked better not just via a java doc.

Or maybe instead of using generic interface, define `RetriableAction` as a static interface in `RetryingExecutor`? I mean this would help to raise awareness of the idempotence requirement and also significantly improved IDE support (looking for usages, implementations etc).